### PR TITLE
Implement NameClass search for `self`

### PR DESCRIPTION
**Pull request by Veykril**
With this Names, Lifetimes and `self` can now be handled similarly.
We also now get hover on `self` values:
Code_kKwxfA6ZvG
And highlighting now gives `self` parameters the `definition` modifier.

### CODE REVIEW

File: crates/ide_db/src/defs.rs [hunk 1]
```diff
@@ -401,6 +408,25 @@ impl NameRefClass {
             _ => None,
         }
     }
+
+    pub fn classify_self_param(
+        sema: &Semantics<RootDatabase>,
+        self_token: &SyntaxToken,
+    ) -> Option<NameRefClass> {
+        assert_eq!(self_token.kind(), syntax::T![self]);
```
Comment (matklad) [inline crates/ide_db/src/defs.rs:416]:
```suggestion
        assert_eq!(self_token.kind(), T![self]);
```

File: crates/ide_db/src/defs.rs [hunk 2]
```diff
@@ -401,6 +408,25 @@ impl NameRefClass {
...
+        let parent = self_token.parent();
+        let segment = ast::PathSegment::cast(parent.clone())?;
+        let path = segment.parent_path();
+        if path.qualifier().is_some() || !ast::PathExpr::can_cast(path.syntax().parent()?.kind()) {
+            // probably a path in a Use-item
+            return None;
+        }
+
+        // FIXME: can we search through macros?
+        let function = parent.ancestors().find_map(ast::Fn::cast)?;
```
Comment (matklad) [inline crates/ide_db/src/defs.rs:426]: `Semantics::ancestors_with_macros`

**Comment by matklad** [comment]
Looking at this: shouldn't we just wrap `crate`, `super` and `self` in `ast::Path` in name ref?

**Comment by Veykril** [comment]
I would believe that NameRef doesn't contain keywords, at least I imagine most things in RA expect that right? I expect this change would simplify this quite a bit though so that might be a good idea. I could test that out later, I guess this change would be a grammar change then if I follow you correctly?

**Comment by matklad** [comment]
NameRef can be a number already (foo.0), so this shouldn't break to much
expectations.

**Comment by Veykril** [comment]
Closing in favor of #7288

**Pull request closed without merging**